### DISCUSSION TITLE
should Link prefetch also fetch the page json?

### DISCUSSION THREAD

**Discussion by ingoclaro** [vercel/next.js#11578] [Help] [answered]
Link component prefetches by default the page javascript
I noticed that when a page implements getStaticProps or getServerSideProps the json data file is only downloaded when actually visiting the page (after click on the link).
I think it would be nice if the Link component could also prefetch this data file together with the page javascript code. Thoughts?

**Comment by timneutkens** [comment, marked as answer]
If the page uses getStaticProps the data is prefetched. When you use getServerSideProps it is not as it would increase server load.

**Reply by lukebarton** [reply]
@timneutkens not if I've brought my own cache. 

To get prefetching, as you say, I need to use `getStaticProps`, and I can use `fallback: true` to get some 'dynamicness', but I don't have control of my cache headers to expire my content 🙃 

What's a man to do if he wants control of his cache headers _and_ prefetching? It doesn't sound unreasonable.

**Reply by uladkasach** [reply]
@timneutkens it'd be great to be able to choose to increased server load in exchange for a better, faster experience for the user. (I.e., opt into prefetching pages which use getServerSideProps)

for example, with a flag on the link like suggested by this discussion: https://github.com/vercel/next.js/discussions/13105

any chance of this happening?

**Reply by robertpiosik** [reply]
@timneutkens up, as seemingly using CDN in front of next was never thought about 👀

**Reply by lukebarton** [reply]
@robertpiosik yep, that's the vibe I get from Next - it's clear to me that they know about CDNs -- their Vercel platform is a one. The fact is that it could be better designed: giving users control over expiry headers on something which is essentially a HTTP response from a server should be out-of-the-box functionality.

**Reply by cvbuelow** [reply]
* https://github.com/vercel/next.js/pull/46210

**Reply by Garrett-R** [reply]
We need this too!  We serve different things from CDN based on a cookie's value so we have to use `getServerSideProps`, but would be nice to get prefetch!

**Comment by ingoclaro** [comment]
I don't see that behavior in firefox 74, onHover only fetches the .js.
However I checked in chrome and indeed it fetches the .js and the .json when the link comes into the viewport when using getStaticProps (this doesn't trigger in firefox).

This in the production build.

**Comment by ingoclaro** [comment]
is the firefox behavior a known and intentional one or a bug?
In summary:
- chrome: fetches .js and .json (if staticProp) when link comes into viewport. Then fetches .js when hovering the link (uses prefetch cache for this one)
- firefox: only fetches .js when hovering the link

**Reply by stefanprobst** [reply]
hi, i'm seeing the exact same behavior, did you figure out why this is not working in firefox?

**Reply by stefanprobst** [reply]
ok, so i had `network.prefetch-next` set to `false` in `about:config` settings - not sure why

**Reply by stefanprobst** [reply]
figured this out, in case anyone else runs into this: "uBlock Origin", one of the popular adblock browser plugins, by default disables prefetching. however, on Chrome this does not seem to do anything, while on Firefox it significantly affects the performance of sites built with next.js. see https://github.com/gorhill/uBlock/wiki/Dashboard:-Settings#disable-pre-fetching

**Comment by valoricDe** [comment]
I think it would be nice to be able to toogle this in next.config.js for getServerSideProps. Because we have a CDN in front of next. So it would only once cause a big load to the server and further requests would be server by the CDN.
Because of the CDN it also doesn't make really sense for us to use getStaticProps because we want to get the newest content when the CDN cache expires

**Reply by fasani-tx** [reply]
On top of that when using a CDN and getServerSide and next/link, the props in the prefetch is an empty {} where as the click and client side hydration is a fully constructed props object. Both have the same URL, I think now I will need to implement a new header on the cache. I am thinking "purpose: prefetch" or "x-middleware-prefetch:1" but not sure which to use for future proofing my app.

**Reply by Fatima-Tahseen** [reply]
@fasani-tx  do we have any official documentation on this change or any reasoning behind this? Also, its happening for me in the pages directory
